doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -443,46 +451,22 @@ the conversion in hypotheses :n:`{+ @ident}`.
    the call-by-value strategy, the latter is generally more efficient for
    evaluating purely computational expressions (i.e. with little dead code).
 
-.. tacv:: compute
-          cbv
-   :name: compute; _
-
-   These are synonyms for ``cbv beta delta iota zeta``.
-
-.. tacv:: lazy
+   .. tacn:: compute {? @delta_flag } {? @occurrences }
 
-   This is a synonym for ``lazy beta delta iota zeta``.
+      A variant form of :tacn:`cbv`.
 
-.. tacv:: compute [ {+ @qualid} ]
-          cbv [ {+ @qualid} ]
+.. tacn:: vm_compute {? {| @reference_occs | @pattern_occs } } {? @occurrences }
 
-   These are synonyms of :n:`cbv beta delta {+ @qualid} iota zeta`.
-
-.. tacv:: compute - [ {+ @qualid} ]
-          cbv - [ {+ @qualid} ]
-
-   These are synonyms of :n:`cbv beta delta -{+ @qualid} iota zeta`.
-
-.. tacv:: lazy [ {+ @qualid} ]
-          lazy - [ {+ @qualid} ]
-
-   These are respectively synonyms of :n:`lazy beta delta {+ @qualid} iota zeta`
-   and :n:`lazy beta delta -{+ @qualid} iota zeta`.
-
-.. tacv:: vm_compute
-   :name: vm_compute
-
-   This tactic evaluates the goal using the optimized call-by-value evaluation
+   Evaluates the goal using the optimized call-by-value evaluation
    bytecode-based virtual machine described in :cite:`CompiledStrongReduction`.
    This algorithm is dramatically more efficient than the algorithm used for the
    :tacn:`cbv` tactic, but it cannot be fine-tuned. It is especially interesting for
    full evaluation of algebraic objects. This includes the case of
    reflection-based tactics.
 
-.. tacv:: native_compute
-   :name: native_compute
+.. tacn:: native_compute {? {| @reference_occs | @pattern_occs } } {? @occurrences }
 
-   This tactic evaluates the goal by compilation to OCaml as described
+   Evaluates the goal by compilation to OCaml as described
    in :cite:`FullReduction`. If Coq is running in native code, it can be
    typically two to five times faster than :tacn:`vm_compute`. Note however that the
    compilation cost is higher, so it is worth using only for intensive
```

Comment:
"run time of the code" and "total size of the code" are not comparable.  Perhaps you would clarify your comment?

Reply:
The conversion factor I'm imagining is the time it takes to run the no-op computation, i.e., to invoke `vm_compute` or `native_compute` on a term for which no reduction remains to be done.  I don't have a good sense of the units of conversion here, but the point I'm making is that what matters for relative efficiency of `native_compute` over `vm_compute` is not just the runtime of the computation, but the ratio of runtime of the computation to runtime of running `vm_compute` as a no-op on all code in the dependency graph of the term being computed.

Reply:
Still not clear.  We're trying to advise when to use one or the other tactic.  I think the advice is to use `vm_compute` up to some threshold and then `native_compute`.  What is a useful rule of thumb?

- `native_compute` time = compile time + run time + ??
- `vm_compute` time = "running `vm_compute` as a no-op on all code in the dependency graph" + run time + ??

`native_compute` run time is 2-5x faster, how do the setup times compare--doesn't have to be precise.  Users will experiment in any case.

Reply:
Useful rule of thumb: if your input code-size is small (you're operating on code you wrote by hand rather than code that came out of a reification procedure or out of tactics generating large terms like many nested destructs on inductives with dozens or hundreds of constructors), *and* your output is small (you're returning a bool or a nat or an N or a Z, etc; rather than e.g., a large AST), *and* the running time is long (at least 5--10 seconds in vm_compute), then native_compute is likely to serve you well.  In all other cases, you want vm_compute.

native_compute time = conversion to native code + compile time + run time + readback time + OS overhead (writing files to disk and linking them)
vm_compute time = bytecode compilation/translation time + run time + readback time

Some relevant stats:
- I believe OS overhead in native is always at least on the order of 0.1s, so you'll never be able to get many small calls to the native compiler to be as fast as the VM
- my vague sense is that readback (turning the optimized representation that the reduction engine operates on back into kernel terms) is somewhat faster for the VM than the native compiler, hence why you don't want to use native when outputting large terms (I haven't extensively profiled this though; I may be wrong)
- conversion to native code is generally pretty fast, I think (perhaps roughly the same as conversion bytecode for the VM?)
- compilation of native code is generally slow, especially on large terms and so if there are no expensive loops that dominate the runtime, and most of the runtime will just be spent unfolding and inlining definitions, the overhead of native compilation won't be worth the modest improvement in runtime.

If you want hard numbers on any of these, I'd need to run some experiments